nx/lib/nx/defn/expr.ex

```diff
@@ -173,13 +173,26 @@ defmodule Nx.Defn.Expr do
   def cond(file, clauses, last) do
     clauses =
       for {meta, {pred, expr}} <- clauses do
-        pred = to_expr(pred)
+        pred =
+          case pred do
+            pred when is_boolean(pred) ->
+              raise CompileError,
+                line: meta[:line],
+                file: file,
+                description: "boolean predicate passed to cond/if, cond/if expects" <>
+                             " scalar tensor predicate, consider using 1 for true" <>
+                             " or 0 for false as an alternative"
+            pred ->
+              to_expr(pred)
+            end
```

Comment:
We have added while now... so maybe we should create a to_pred function and encapsulate this there? Then call it from cond and where.